src/Runtime/Distributions/Automata/MultiRepresentationWeightFunction.cs

```diff
@@ -0,0 +1,758 @@
+    /// <typeparam name="TAutomaton">The type of a weighted finite state automaton, that can be used to
+    /// represent all weight functions.</typeparam>
+    [Serializable]
+    [DataContract]
+    [Quality(QualityBand.Experimental)]
+    public struct MultiRepresentationWeightFunction<TSequence, TElement, TElementDistribution, TSequenceManipulator, TPointMass, TDictionary, TAutomaton> :
+        IWeightFunction<TSequence, TElement, TElementDistribution, TSequenceManipulator, TAutomaton, MultiRepresentationWeightFunction<TSequence, TElement, TElementDistribution, TSequenceManipulator, TPointMass, TDictionary, TAutomaton>>
+        where TSequence : class, IEnumerable<TElement>
+        where TElementDistribution : IDistribution<TElement>, SettableToProduct<TElementDistribution>, SettableToWeightedSumExact<TElementDistribution>, CanGetLogAverageOf<TElementDistribution>, SettableToPartialUniform<TElementDistribution>, new()
+        where TSequenceManipulator : ISequenceManipulator<TSequence, TElement>, new()
+        where TPointMass : PointMassWeightFunction<TSequence, TElement, TElementDistribution, TSequenceManipulator, TAutomaton, TPointMass>, new()
+        where TDictionary : DictionaryWeightFunction<TSequence, TElement, TElementDistribution, TSequenceManipulator, TAutomaton, TDictionary>, new()
+        where TAutomaton : Automaton<TSequence, TElement, TElementDistribution, TSequenceManipulator, TAutomaton>, new()
+    {
+        private static TSequenceManipulator SequenceManipulator =>
+                Automaton<TSequence, TElement, TElementDistribution, TSequenceManipulator, TAutomaton>.SequenceManipulator;
+
+        private static readonly TAutomaton ZeroAutomaton =
+            Automaton<TSequence, TElement, TElementDistribution, TSequenceManipulator, TAutomaton>.Zero();
+
+        private const int MaxDictionarySize = 128;
+
+        /// <summary>
+        /// A function mapping sequences to weights.
+        /// Can only be of one of the following types: TPointMass, TDictionary, TAutomaton, or <see langword="null"/>.
+        /// <see langword="null"/> should be interpreted as zero function.
+        /// </summary>
+        [DataMember]
+        private IWeightFunction<TSequence, TElement, TElementDistribution, TSequenceManipulator, TAutomaton> weightFunction;
+
+        /// <summary>
+        /// An <see cref="IWeightFunctionFactory{TSequence, TElement, TElementDistribution, TSequenceManipulator, TAutomaton, TWeightFunction}"/>
+        /// implementation for <see cref="MultiRepresentationWeightFunction{TSequence, TElement, TElementDistribution, TSequenceManipulator, TPointMass, TDictionary, TAutomaton}"/>.
+        /// </summary>
+        public class Factory : IWeightFunctionFactory<TSequence, TElement, TElementDistribution, TSequenceManipulator, TAutomaton, MultiRepresentationWeightFunction<TSequence, TElement, TElementDistribution, TSequenceManipulator, TPointMass, TDictionary, TAutomaton>>
```

Comment:
While they indeed look terrible, I'd argue, that that can of worms was already opened by Automaton, SequenceDistribution, and related types. Unfortunately, it isn't possible to hide the full MultiRepresentationWeightFunction<...> by inheriting a closed constructed version of it, because it's a struct. As for suggestions:
1. I'll have to think on that.
1. No, dictionaries and automata make actual use of inheritance, so the types have to be specified.

Reply:
I think inheritance is used only to minimize typing. (I.e. to give a name to the huge parametrieze thing). If we move things into a class, then we could probably fully avoid it.

Reply:
No, there're method overloads in StringDictionaryWeightFunction and StringAutomaton.

Reply:
OK, I implemented 1, it indeed seems to have simplified the code somewhat. Also, removed unnecessary `TThis` from `PointMassWeightFunction` and sealed it instead, so no need to specify this one for `MultiRepresentableWeightFunction` anymore. It still is parametrized by a type for dictionary weight function, though.

Had to fix a todo in compiler for this to work - https://github.com/dotnet/infer/pull/339/commits/eb0dfcea4a083a3a5a58a10d8ac86d0681f1c00b